# TME 3 :  Evaluation des résultats de  méthode de Prédiction de gènes



**Nom etudiant 1 : K Murali Sharane**

**Nom etudiant 2 : Legendre--Despas Jeanne**

### A) Matrice de confusion et mesures de performance 

Il est nécessaire d’avoir une méthode pour évaluer la qualité des prédictions obtenues pour les coordonnées des gènes. Est ce que certains gènes sont ratés par la prédiction, ou est-ce qu’au contraire certains des gènes sont systématiquement surannotés alors qu’il s’agit d’une région intergénique (entre les gènes) ? Toute l’information peut être résumée en construisant le tableau de valeurs suivante, appelée matrice de confusion. Pour cela nous attribuons à chaque nucléotide d’un génome donné la valeur 0 ou 1, 0 si le nucléotide ne fait pas partie d’un gène et 1 si il en fait partie. La matrice de confusion est calculée comme ce tableau. 

![](matriceconfusion.png "")

On résume ensuite la qualité de l’annotation avec les métriques suivantes : 

   + La sensibilité  (= ``11/(11+10)``) <br>
Sen = (le nombre de vrais positifs) / (le nombre de vrais positifs + le nombre de faux négatifs)

   + La spécificité (= ``00 / (00+01)``) <br>
Spe = (le nombre de vrais négatifs) / (le nombre de vrais négatifs + le nombre de faux positifs)

   + La valeur prédictive VP (= ``11 / (11+01)``) est le taux de vrais positifs parmi ceux prédits comme positifs <br>
VP = (le nombre de vrais positifs) / (le nombre de vrais positifs + le nombre de faux positifs)


**Nous allons d'abord coder plusieurs fonctions qui seront utilisées plus tard.**




<b>Question 1)</b> Ecrivez une fonction ``eval_res`` qui calcule la **Sen**, la **Spe** et **VP** (la valeur prédictive) à partir d’une matrice de confusion et renvoie un tuple à trois dimensions. La matrice de confusion sera codée comme un tableau à deux dimension, _e.g._ une liste de listes ou une matrice de numpy.


In [4]:
import doctest # Pour pouvoir utiliser doctest.testmod()
import numpy as np

In [5]:
# Question 1
def eval_res(matconf):
    """Renvoi un tuple avec la sensibilité, la spécificité et la valeur prédictivindices_off = [i.start() for i in re.finditer(fauxNegOff, str_intervalle_off)]
indices_tme2 = [i.start() for i in re.finditer(fauxNegTme, str_intervalle_tme2)]e.
    entrée1 matconf : matrice de confusion (les positions contient VP, FP, FN, VN) 
    sortie1 sen: sensibilité 
    sortie2 spe: spécificité 
    sortie3 vp: sensibi valeur prédictive 
    >>> eval_res([[4, 1], [1, 3]])
    (0.75, 0.8, 0.75)
    """
    
    sen = matconf[1][1]/(matconf[1][1] + matconf[1][0])
    spe = matconf[0][0]/(matconf[0][0] + matconf[0][1])
    vp = matconf[1][1]/(matconf[1][1] + matconf[0][1])
    return (sen, spe, vp)
doctest.testmod()

TestResults(failed=0, attempted=2)

In [6]:
assert eval_res([[50, 10], [5, 100]]) == (100 / 105, 50 / 60, 100 / 110)



<b>Question 2)</b> Ecrivez une fonction ``ecrit_intervalle`` qui, à partir des deux listes des positions de début et de fin des gènes sur le brin sens et de la longueur du génome, recode une grande liste avec la valeur 1 pour les positions qui sont des gènes et 0 sinon. Attention, il faut tenir compte des cas possibles où des gènes se chevauchent. Vous pourrez par exemple initialiser une liste avec ``lg`` valeurs à ``0`` et traiter ensuite les intervalles séquentiellement.


In [7]:
# Question 2
def ecrit_intervalle(positions_debut, positions_fin, len_genome):
    """
    Renvoie une liste binaire avec un élément par base du génome.
    1 indique la présence d'un gène.
    entrée1 positions_debut : liste contenant les positions de debut des genes
    entrée2 positions_fin : liste contenant les positions de fin des genes
    entrée3 len_genome : taille du genome
    sortie1 genome_list : Liste binaire avec un élément par base du génome
    

    >>> ecrit_intervalle([2, 10, 15], [7, 12, 20], 22)
    [0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0]
    """
    genome_list = [0 for _ in range(len_genome)] #initialise tous les positions à zero
    
    for i in range(len(positions_debut)) : 
        for j in range(positions_debut[i], positions_fin[i]+1) :
            genome_list[j] = 1
   
    return genome_list


doctest.testmod()

TestResults(failed=0, attempted=2)

**Question 3)** Ecrivez la fonction ``compare_intervalles`` qui compare deux listes produites par la fonction ``ecrit_intervalle``, et renvoie la matrice de confusion.  
    Par exemple, si les deux listes données en paramètre sont  
    ``génome =[0,0,1,1,1,1,1,1,0,0,1,1,1,0,0,1,1,1,1,1,1,0]`` et  
    ``ORFs   =[0,0,0,1,1,1,1,0,0,0,1,1,1,0,0,0,0,1,1,1,1,0]``, on a :  

| Genome / ORF | 0 | 1 |
|---|---|---| 
| **0** | 7 | 0  |
| **1** | 4 | 11 |


In [8]:
# Question 3
def compare_intervalles(intervalle_1, intervalle_2):
    """Renvoi la matrice de confusion.
    entrée1 intervalle_1: Liste binaire representant un genome de reference
    entrée2 intervalle_2: Liste binaire representant les predictions de genes
    sortie1 matrix: matrice de confusion
    >>> compare_intervalles([0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0], [0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0])
    [[7, 0], [4, 11]]
    """
    matrix = [[0, 0], [0, 0]]
    
    for i in range(len(intervalle_1)) : 
        if (intervalle_1[i] and intervalle_2[i]) :
            matrix[1][1] += 1
        if (intervalle_1[i] and not intervalle_2[i]) :
            matrix[1][0] += 1
        if (not intervalle_1[i] and intervalle_2[i]) :
            matrix[0][1] += 1
        if (not intervalle_1[i] and not intervalle_2[i]) :
            matrix[0][0] += 1

    return matrix

In [9]:
compare_intervalles(
    [0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0],
    [0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0])

[[7, 0], [4, 11]]

### B) Evaluation sur _E. coli_
Nous allons comparer l’annotation des gènes de E. coli disponible au NCBI avec la prédiction par les ORFs realisé au TME precedent.
**Attention, certains gènes définit dans les fichiers .tab sont non-codant (rRNA transfer, rRNA ribosomal). Ces gènes ne commencent pas par un codon START et nous allons les ignorer dans notre comparaison**. Nous allons donc travailler seulement avec les lignes "protein_coding" et le strand "+". 






Nous allons utiliser pandas, une structuré de données puissante qui va simplifier nos analyses.

Pour apprendre les bases on vous conseil les tutoriaux et vidéos suivant

*   https://eric.univ-lyon2.fr/~ricco/tanagra/fichiers/fr_Tanagra_Data_Manipulation_Pandas.pdf
*   https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html (doc en anglais)
*   https://www.youtube.com/watch?v=zZkNOdBWgFQ&t=187s (video en français)



In [11]:
import pandas as pd
import gzip
import shutil
import urllib.request


<b>Preparation :</b> Telecharger le fichier feature_table de E. coli qui contient l'annotation de son genome

In [112]:
#Telecharger le fichier feature_table de E. coli qui contient l'annotation de son genome
urllib.request.urlretrieve('ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/026/345/GCF_000026345.1_ASM2634v1/GCF_000026345.1_ASM2634v1_feature_table.txt.gz',
    'GCF_000026345.1_ASM2634v1_feature_table.txt.gz')

KeyboardInterrupt: 

In [113]:
#Decompresser le fichier
with gzip.open('GCF_000026345.1_ASM2634v1_feature_table.txt.gz', 'rb') as f_in:
    with open('GCF_000026345.1_ASM2634v1_feature_table.txt', 'wb') as f_out:
        shutil.copyfileobj(f_in, f_out)

**Question 1)**: Lire le fichier `GCF_000026345.1_ASM2634v1_feature_table.txt` à l'aide de Pandas

In [12]:
#Lire le fichier à l'aide de Pandas
#pd.read_table() est équivalent
feature_table = pd.read_csv('GCF_000026345.1_ASM2634v1_feature_table.txt', sep = '\t')
print(feature_table.shape)


(9896, 20)


In [13]:
#afficher les premières 5 lignes
feature_table.head()


# feature           class         assembly     assembly_unit    seq_type  \
0      gene  protein_coding  GCF_000026345.1  Primary Assembly  chromosome   
1       CDS    with_protein  GCF_000026345.1  Primary Assembly  chromosome   
2      gene  protein_coding  GCF_000026345.1  Primary Assembly  chromosome   
3       CDS    with_protein  GCF_000026345.1  Primary Assembly  chromosome   
4      gene  protein_coding  GCF_000026345.1  Primary Assembly  chromosome   

   chromosome genomic_accession  start   end strand product_accession  \
0         NaN       NC_011750.1    189   254      +               NaN   
1         NaN       NC_011750.1    189   254      +    WP_001386572.1   
2         NaN       NC_011750.1    335  2797      +               NaN   
3         NaN       NC_011750.1    335  2797      +    WP_001264707.1   
4         NaN       NC_011750.1   2799  3731      +               NaN   

  non-redundant_refseq  related_accession  \
0                  NaN                NaN   
1       WP_001386572.1                NaN   
2                  NaN                NaN   
3       WP_001264707.1                NaN   
4                  NaN                NaN   

                                                name symbol  GeneID  \
0                                                NaN   thrL     NaN   
1                          thr operon leader peptide   thrL     NaN   
2                                                NaN   thrA     NaN   
3  bifunctional aspartate kinase/homoserine dehyd...   thrA     NaN   
4                                                NaN   thrB     NaN   

         locus_tag  feature_interval_length  product_length  \
0  ECIAI39_RS00005                       66             NaN   
1  ECIAI39_RS00005                       66            21.0   
2  ECIAI39_RS00010                     2463             NaN   
3  ECIAI39_RS00010                     2463           820.0   
4  ECIAI39_RS00015                      933             NaN   

                   attributes  
0  old_locus_tag=ECIAI39_4936  
1                         NaN  
2  old_locus_tag=ECIAI39_0001  
3                         NaN  
4  old_locus_tag=ECIAI39_0002

**Question 2)**: Filtrer feature table et laisser seulement les lignes contenant "protein_coding" et le strand "+".

In [14]:
orf_genome = feature_table.loc[(feature_table['class'] == "protein_coding") & (feature_table['strand'] == "+"),:]
orf_genome.head()

# feature           class         assembly     assembly_unit    seq_type  \
0      gene  protein_coding  GCF_000026345.1  Primary Assembly  chromosome   
2      gene  protein_coding  GCF_000026345.1  Primary Assembly  chromosome   
4      gene  protein_coding  GCF_000026345.1  Primary Assembly  chromosome   
6      gene  protein_coding  GCF_000026345.1  Primary Assembly  chromosome   
8      gene  protein_coding  GCF_000026345.1  Primary Assembly  chromosome   

   chromosome genomic_accession  start   end strand product_accession  \
0         NaN       NC_011750.1    189   254      +               NaN   
2         NaN       NC_011750.1    335  2797      +               NaN   
4         NaN       NC_011750.1   2799  3731      +               NaN   
6         NaN       NC_011750.1   3732  5018      +               NaN   
8         NaN       NC_011750.1   5232  5528      +               NaN   

  non-redundant_refseq  related_accession name symbol  GeneID  \
0                  NaN                NaN  NaN   thrL     NaN   
2                  NaN                NaN  NaN   thrA     NaN   
4                  NaN                NaN  NaN   thrB     NaN   
6                  NaN                NaN  NaN   thrC     NaN   
8                  NaN                NaN  NaN    NaN     NaN   

         locus_tag  feature_interval_length  product_length  \
0  ECIAI39_RS00005                       66             NaN   
2  ECIAI39_RS00010                     2463             NaN   
4  ECIAI39_RS00015                      933             NaN   
6  ECIAI39_RS00020                     1287             NaN   
8  ECIAI39_RS00025                      297             NaN   

                   attributes  
0  old_locus_tag=ECIAI39_4936  
2  old_locus_tag=ECIAI39_0001  
4  old_locus_tag=ECIAI39_0002  
6  old_locus_tag=ECIAI39_0003  
8  old_locus_tag=ECIAI39_0004

**Question 3)**: Lire le fichier annotation_des_orfs.csv (produit au TME2) à l'aide de Pandas

In [15]:
orf_tme2 = pd.read_csv('annotation_des_orfs_prof.csv', sep = '\t')
orf_tme2.head()


start   len                                                seq
0    292  2505  GTGCGGGCTTTTTTTTCGACCAAAGGTAACGAGGTAACAACCATGC...
1   2798   933  ATGGTTAAAGTTTATGCCCCGGCTTCCAGTGCCAATATGAGCGTCG...
2   3731  1287  ATGAAACTCTACAATCTGAAAGATCACAATGAGCAGGTCAGCTTTG...
3   8272   804  TTGATTGATGATGCTGTCGCCTGGGCGAAACAGCAGAGCAACGATC...
4  12100  1863  ATGGATGGCACCACTCCTCGTGTACTGGAGAACGCCGAAGGCGATC...

**Question 4)**: Appliquez la fonction ``compare_intervalle`` à l’annotation des gènes sur le brin sens et à la prédiction par les ORFs sur le brin sens,  puis générer la matrice de confusion.

In [16]:
len_genome = 5132068

#Récupération des variables sur l'annotation officielle du génome
positions_debut_off = [k - 1 for k in orf_genome['start'].tolist()]
positions_fin_off = [k - 1 for k in orf_genome['end'].tolist()]

#Récupération des variables sur l'annotation du tme2 du génome
positions_debut_tme2 = orf_tme2['start'].tolist()
len_tme2 = orf_tme2['len'].tolist()
positions_fin_tme2 = [positions_debut_tme2[k] + len_tme2[k] - 1 for k in range(len(positions_debut_tme2))]

#Écriture des intervalles
intervalle_off = ecrit_intervalle(positions_debut_off, positions_fin_off, len_genome)
intervalle_tme2 = ecrit_intervalle(positions_debut_tme2, positions_fin_tme2, len_genome)

In [17]:
#Comparaison des intervalles
matrix = compare_intervalles(intervalle_off, intervalle_tme2)

In [18]:
matrix

[[2832085, 289576], [438702, 1571705]]

matrix = [[2832085, 289576], [438702, 1571705]]

**Question 5)**: Appliquez la fonction ``eval_res`` à la matrice de confusion et affichez Sen, Spe et VP.

In [19]:
var = eval_res(matrix)
print("Sensibilité : ", var[0])
print("Spécificité : ", var[1])
print("Valeur prédictive : ", var[2])


Sensibilité :  0.7817844844352412
Spécificité :  0.9072365641240352
Valeur prédictive :  0.8444211271699437


<b>Question 6)</b>: Analyse des résultats. Pour comprendre les limitations d'une méthode de prédiction on doit toujours analyser dans un second temps les erreurs faites par le programme. En détectant des erreurs récurrente, il est plus facile de proposer ensuite des améliorations. Donnez deux exemples de régions d'au moins 50 nucléotides consécutifs qui sont :

*   un faux négatif (gène raté)
*   un faux positif (gène inexistant)

A votre avis à quoi ces deux types d'erreurs sont-elles dues ? 

In [20]:
import re

In [21]:
str_intervalle_off = "".join(str(elem) for elem in intervalle_off)
str_intervalle_tme2 = "".join(str(elem) for elem in intervalle_tme2)
seq1 = 50 * "1"
seq0 = 50 * "0"


indices_off_fn = [i.start() for i in re.finditer(seq1, str_intervalle_off)]
indices_tme2_fn = [i.start() for i in re.finditer(seq0, str_intervalle_tme2)]

In [26]:
def faux_neg():
    for i_off in indices_off_fn :
        for i_tme2 in indices_tme2_fn :
            if i_off == i_tme2 :
                return i_off

faux_neg = faux_neg()      
print(faux_neg)

52385


In [27]:
print(str_intervalle_off[faux_neg:faux_neg+50])
print(str_intervalle_tme2[faux_neg:faux_neg+50])

11111111111111111111111111111111111111111111111111
00000000000000000000000000000000000000000000000000


In [28]:
indices_off_fp = [i.start() for i in re.finditer(seq0, str_intervalle_off)]
indices_tme2_fp = [i.start() for i in re.finditer(seq1, str_intervalle_tme2)]

In [30]:
def faux_pos() :
    for i_off in indices_off_fp :
        for i_tme2 in indices_tme2_fp :
            if i_off == i_tme2 :
                return i_off
faux_pos = faux_pos()
print(faux_pos)

323907


In [31]:
print(str_intervalle_off[faux_pos:faux_pos+50])
print(str_intervalle_tme2[faux_pos:faux_pos+50])

00000000000000000000000000000000000000000000000000
11111111111111111111111111111111111111111111111111


Notre fonctions parcourt une seule fois le génome en cherchant un start tous cadre de lecture confondu puis un stop dans le même cadre de lecture et ainsi de suite. 
Ainsi les gène chevauchant ou nichés ne sont pas ou mal detecté. Ce qui peuvent entrainer des erreurs de faux négatifs. Cela peut aussi décaller la détection des ORFs et mener à la détection d'ORFs qui n'en sont pas, soit des faux négatifs.
Les faux négatifs et les faux positif peuvent être également dus à la présence d'introns.
Ces introns peuvent être perçus comme des séquences codantes par notre algorithmes et dont créer des faux positifs. Ils peuvent également contenir des codons STOPs car ils ne sont pas soumis à des pressions de sélection et ainsi venir terminer précocement nos ORfs. Ces gènes ne sont alors plus détectés et seront des faux négatifs pour notre algorithme.